src/components/AutoComplete/AsyncAutoComplete.js

```diff
@@ -3,34 +3,47 @@ import { AsyncTypeahead } from 'react-bootstrap-typeahead';
 import PropTypes from 'prop-types';
 
 class AsyncAutoComplete extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {
-      options: [],
-      isLoading: false
-    };
-  }
+  state = {
+    options: this.props.options,
+    isLoading: this.props.isLoading
+  };
+
+  finishCallback = options => this.setState({ options, isLoading: false });
 
   handleSearch = () => {
     this.setState({ isLoading: true });
-    const options = this.props.onSearch() || [];
-    this.setState({ options, isLoading: false });
+    // The user must pass the callback to
+    // the 'onSearch' function in order to update the state properly.
+    this.props.onSearch(this.finishCallback);
```

Comment:
This is actually a react antipatern.  To handle loading data async will normally follow a pattern similar to:
```jsx
class SomeConsumer extends React.Component {
  state = {
    isLoading: false,
    options: [],
  };

  handleSearch = (query) => {
    if (this.state.isLoading) {
      return;
    }

    this.setState({ isLoading: true });
    return fetch('https://api.github.com/repos/patternfly/patternfly-react/forks')
      .then(response => response.json())
      .then(forks => {
        this.setState({
          options: forks.filter(fork => fork.owner.includes(query)),
          isLoading: false,
        });
      });
  }

  render() {
    return (
      <AsyncTypeahead
        isLoading={this.state.isLoading} 
        onSearch={this.state.onSearch}
        options={this.state.options}
      />
    );
  }
}
```

With this pattern,  we actually don't need this wrapper (AsyncAutoComplete). We could probably just re-export `AsyncTypeahead` instead of wrapping it, and set the default props.

Reply:
hey @dmiller9911, The wrapper allows us to use this component without the need of state,
such as in Storybook, providing only the async function. 
I think it might be useful in some scenarios

Reply:
I can see the use case of it.  I think if we are going to allow it we should make it a promise instead of using a callback.  it will be easier to follow.  We can actually support both promises and regular values by doing `Promise.resolve(this.props.onSearch({ query })).then(options => ...)`.  I think it might also be useful to make query part of an object that we pass to search.  This will allow us to send more data in the future without needing a breaking change or adding more params.

Reply:
This is a great idea @dmiller9911 ! much easier for the developer to set, 
although I think we should leave the `query`  to be implemented by the developers
and deal only with the options they delivered.

For example:
`<AsyncAutoComplete
        {...props}
        onSearch={ () =>
           fetch('https://api.github.com/repos/patternfly/patternfly-react/forks')
              .then(response => response.json())
              .then(forks => forks.map(fork => fork.owner))
        }
 />`

in AsyncAutoComplete.js:
`handleSearch = () => {
    this.setState({ isLoading: true });
    Promise.resolve(this.props.onSearch()).then(options =>
      this.finishSearch(options)
    );
  };`